elastic/beats
# Fix terminating pod autodiscover issue

## What does this PR do?
This addition handles the case when a Pod is in `Terminating` phase. 
In this case the pod is neither `PodSucceeded` nor `PodFailed` and
hence requires special handling.
In this, we check all Pod's containers' statuses and if we find a at least
one still running we ignore this update event. With this, we make sure
that all containers are stopped before stopping Filebeat.


## Why is it important?
To collect any leftover logs when the pod is going in Terminating state.

## Checklist

~~- [ ] I have made corresponding changes to the documentation~~
~~- [ ] I have made corresponding change to the default configuration files~~
~~- [ ] I have added tests that prove my fix is effective or that my feature works~~

## How to test this PR locally
Use replication steps of https://github.com/elastic/beats/issues/17396#issue-591883875 and expect to see that the issue is fixed. The screenshot below provides the fixed case.


```
filebeat.autodiscover:
      providers:
        - type: kubernetes
          node: ${NODE_NAME}
          templates:
            - condition:
                equals:
                  kubernetes.pod.name: "mytargetpod"
              config:
                - type: container
                  paths:
                    - /var/log/containers/*${data.kubernetes.container.id}.log
```

## Related issues

- Closes https://github.com/elastic/beats/issues/17396
- Relates https://github.com/elastic/beats/pull/14259



## Screenshots
![Screenshot 2020-07-21 at 12 17 45](https://user-images.githubusercontent.com/11754898/88037866-1fd28780-cb4e-11ea-94d8-9a8e114f2f49.png)
